🐛  Fix invalid image URLs not being cached and causing timeouts

aileen opened the pull request:
refs #8868

- swapped the usage of `got` for requests with the request util, which validates a URL before doing a request and resolving the Promises in image-size so it gets saved in the image-size cache.
- Uses catch predicates instead of conditionals in `getImageSizeFromUrl`
- Return `NotFoundError` if applicable in `getImageSizeFromFilePath` as the caller function `cachedImageSizeFromUrl` is differentiating those between this error and others.
- Using `ImageObject` as a global var resulted in having the `url` property being the same for all requests coming in.
- The logic that checked for an existing protocol (e. g. gravatar URLs) was overly complicated. Refactored it to be more simple.
- Passing the correct value to `fetchDimensionsFromBuffer` as the population of `imageObject.url` happens there. These are used in our structured data and need to be full URLs (in case of locally stored files) or the original URL (in case of URLs missing the protocol)
- Added two more debug logs in `getCachedImageSizeFromUrl` so it's logged when an image is added to the cache even tho it was returned as error.
- We're always resolving the result in `getCachedImageSizeFromUrl`, so there's no need to return the values with a `Promise.resolve()`. The caller fn uses waits for the Promises to be fulfilled.

Todos:
- [x] investigate and refactor usage of `Promise.resolve()` in cached image size
- [x] Use predicate catches in image-size
- [x] Add `NotFoundError` handling to storage.read request, as the calling fn `cachedImageSizeFromUrl` is differentiating those errors
- [x] check if global var `imageObject` interferes with image size cache
- [x] find and improve unnecessary code (nice to have)
- [x] solve catch predicates issue see https://github.com/TryGhost/Ghost/pull/8986#issuecomment-327992639

aileen committed:
📐  Use request util in image-size

refs #8868

- swapped the usage of `got` for requests with the request util

aileen committed:
💄  Use catch predicates

refs #8868

- Uses catch predicates instead of conditionals in `getImageSizeFromUrl`
- Return `NotFoundError` if applicable in `getImageSizeFromFilePath` as the caller function `cachedImageSizeFromUrl` is differentiating those between this error and others.

AileenCGN commented:
@ErisDS one question I have on my mind is the handling of the different errors in `cachedImageSizeFromUrl`. Right now, we're not logging or producing the `NotFoundErrors` see [here](https://github.com/TryGhost/Ghost/blob/master/core/server/utils/cached-image-size-from-url.js#L29) but everything else and I'm wondering why? Mostly the question if timeout errors should also be caught here, as they can be caused by a redirect which is indeed a bit slower now.

aileen committed:
🐛  Fixed ImageObject URL & simplify no protocol URL logic

- Using `ImageObject` as a global var resulted in having the `url` property being the same for all requests coming in.
- The logic that checked for an existing protocol (e. g. gravatar URLs) was overly complicated. Refactored it to be more simple.
- Passing the correct value to `fetchDimensionsFromBuffer` as the population of `imageObject.url` happens there. These are used in our structured data and need to be full URLs (in case of locally stored files) or the original URL (in case of URLs missing the protocol)
- Added two more debug logs in `getCachedImageSizeFromUrl` so it's logged when an image is added to the cache even tho it was returned as error.

aileen committed:
👀  Differentiate error codes between request and storage

AileenCGN commented on core/server/utils/image-size.js:
```diff
@@ -165,27 +170,43 @@ getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
+                message: err.message,
+                code: 'IMAGE_SIZE',
+                err: err,
+                context: {
```
@ErisDS I don't know if this makes sense, but I thought it would be useful information to provide the passed in filepath along with the path we used to read from the storage?

ErisDS replied:
`context` is for users and must be a string. If you want to provide extra details along with the error for debugging purposes, please use `errorDetails: {}`

AileenCGN commented:
There seems to be still some issues with the errors returned from the `got`, is the catch predicates in `getCachedImageSizeFromUrl` don't handle it properly. It works totally fine with errors returned from reading the storage. 

So here's what's happening:

Any filtered catch (e. g. `NotFoundError`) works in `getImageSizeFromUrl` but then it seems like _this_ error (at least it is not the `got` error anymore, but the `NotFoundError` we returned) gets caught by the 'catch all' predicate and that's what get's returned to `getCachedImageSizeFromUrl` in the end so the predicates there won't work. I can't explain why this happens. It works fine when I remove the 'catch all' predicate in `getImageSizeFromUrl`, but this is definitely not a solution!

aileen committed:
🔥  Remove not needed `Promise.resolve()`

We're always resolving the result in `getCachedImageSizeFromUrl`, so there's no need to return the values with a `Promise.resolve()`. The caller fn uses waits for the Promises to be fulfilled.

ErisDS commented:
> @ErisDS one question I have on my mind is the handling of the different errors in cachedImageSizeFromUrl. Right now, we're not logging or producing the NotFoundErrors see here but everything else and I'm wondering why? Mostly the question if timeout errors should also be caught here, as they can be caused by a redirect which is indeed a bit slower now.

We don't do anything with not found errors because they are an expected case. Images will 404 over time. We don't want to fill up the logs, nothing went wrong :)

> Any filtered catch (e. g. NotFoundError) works in getImageSizeFromUrl but then it seems like this error (at least it is not the got error anymore, but the NotFoundError we returned) gets caught by the 'catch all' predicate and that's what get's returned to getCachedImageSizeFromUrl in the end so the predicates there won't work. I can't explain why this happens. It works fine when I remove the 'catch all' predicate in getImageSizeFromUrl, but this is definitely not a solution!

This is expected behaviour as far as I know. If you catch an error in a predicate and then re-throw (reject) it, it will next get caught by the catch all. The way to deal with this is to check in the catch-all for any errors which are not already GhostErrors. If not, wrap them, else, just reject again.

AileenCGN commented:
>We don't do anything with not found errors because they are an expected case. Images will 404 over time. We don't want to fill up the logs, nothing went wrong :) 

I am sorry, I expressed myself wrong :( I was referring to logging the errors. My concern was only, that also timeout errors might now increase (because of the ability to follow redirects and they actually take a bit longer and might hit the timeout) and will therefore be logged. My question was, if we want them to be logged as well?

ErisDS commented:
@AileenCGN sounds like you might be onto something - my advice would be to experiment with what feels right.

aileen committed:
☂️  Wrap already rejected predicate errors in catch all

AileenCGN commented:
@ErisDS This PR is ready for review and merge 🎉

ErisDS commented on core/server/utils/image-size.js:
```diff
@@ -143,6 +143,9 @@ getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
             context: err.url || imagePath
         }));
     }).catch(function (err) {
+        if (err instanceof errors.NotFoundError || err instanceof errors.InternalServerError) {
```
This can be simplified to `err instanceof errors.GhostError` I think - cc @kirrg001 on the proper way to detect pre-handled errors (we already do this here though: https://github.com/TryGhost/Ghost/blob/master/core/server/helpers/register.js#L20)

We should trust that if the error has already been handled it was handled correctly.

AileenCGN replied:
Oops.... I thought I tried that at first and it didn't work, but I didn't do it right. Pushed the changes.

ErisDS commented on core/server/utils/image-size.js:
```diff
@@ -199,6 +202,9 @@ getImageSizeFromFilePath = function getImageSizeFromFilePath(imagePath) {
                 }
             }));
         }).catch(function (err) {
+            if (err instanceof errors.NotFoundError || err instanceof errors.InternalServerError) {
```
Same here - no need to look for specific types of GhostError

ErisDS commented:
Just checking: as a result of these changes, what now happens in the original error case I had with the old-nonsubdirectory local path?

AileenCGN commented:
> Just checking: as a result of these changes, what now happens in the original error case I had with the old-nonsubdirectory local path?

This case will result in an `invalid or missing URL` error thrown [here](https://github.com/TryGhost/Ghost/blob/master/core/server/utils/request.js#L9) and will be caught here: https://github.com/TryGhost/Ghost/pull/8986/files#diff-6d3ba398735146e7c9d8c64ac57dbc5bR34

The cache issue is resolved as well, so a second request results in reading from the cache.

aileen committed:
Use errorDetails instead of context

ErisDS commented:
Awesome 👍 

There are a couple of minor amends needed. Will review again first thing in the am :)

AileenCGN commented:
Summary of cases I successfully tested:

- Local storage files
    - Missing local image
    - old non-subdirectory path
- Image URL
    - redirected image
    - invalid image URL
    - image not found (404)
    - timeouts for image requests

Of course I tested positive cases as well ;) 
For all of these cases, the cache works and the error handling does its job as well.

aileen committed:
☂️  Support /assets/ image paths

closes TryGhost/Support#361

- adds a guard that checks the image URL for `/assets/` in the beginning and passes a completed URL to the request util to try and fetch the image size
- adds tests

AileenCGN commented:
@ErisDS fyi I left commits separate to make the review easier, but they should totally be squashed when being merged.
